fsspec/implementations/smb.py

```diff
@@ -0,0 +1,225 @@
+
+    @classmethod
+    def _share_has_path(cls, path):
+        parts = path.count("/")
+        if path.endswith("/"):
+            return parts > 2
+        return parts > 1
+
+    @classmethod
+    def _strip_protocol(cls, path):
+        return infer_storage_options(path)["path"]
+
+    @staticmethod
+    def _get_kwargs_from_urls(urlpath):
+        # smb://workgroup;user:password@host:port/share/folder/file.csv
+        out = infer_storage_options(urlpath)
+        out.pop("path", None)
+        out.pop("protocol", None)
+        return out
+
+    def mkdir(self, path, create_parents=True, **kwargs):
+        wpath = self._as_unc_path(path)
+        if create_parents:
+            smbclient.makedirs(wpath, exist_ok=False, **kwargs)
+        else:
+            smbclient.mkdir(wpath, **kwargs)
+
+    def makedirs(self, path, exist_ok=False):
+        if self._share_has_path(path):
+            wpath = self._as_unc_path(path)
+            smbclient.makedirs(wpath, exist_ok=exist_ok)
+
+    def rmdir(self, path):
+        if self._share_has_path(path):
+            wpath = self._as_unc_path(path)
+            smbclient.rmdir(wpath)
+
+    def info(self, path):
+        wpath = self._as_unc_path(path)
+        s = smbclient.stat(wpath)
+        if S_ISDIR(s.st_mode):
+            t = "directory"
+        elif S_ISLNK(s.st_mode):
+            t = "link"
+        else:
+            t = "file"
+        res = {
+            "name": path + "/" if t == "directory" else path,
+            "size": s.st_size,
+            "type": t,
+            "uid": s.st_uid,
+            "gui": s.st_gid,
+            "time": s.st_atime,
+            "mtime": s.st_mtime,
+        }
+        return res
+
+    def ls(self, path, detail=False):
+        unc = self._as_unc_path(path)
+        dirs = ["/".join([path.rstrip("/"), p]) for p in smbclient.listdir(unc)]
+        if detail:
+            dirs = [self.info(d) for d in dirs]
+        return dirs
+
+    # def put(self, lpath, rpath):
+    #     self.ftp.put(lpath, rpath)
+
+    # def get(self, rpath, lpath):
+    #     self.ftp.get(rpath, lpath)
+
+    def _open(
+        self,
+        path,
+        mode="rb",
+        block_size=-1,
+        autocommit=True,
+        cache_options=None,
+        **kwargs
+    ):
+        """
+        block_size: int or None
+            If 0, no buffering, if 1, line buffering, if >1, buffer that many
+            bytes, if None use default from paramiko.
+        """
+        buffering = block_size if block_size is not None and block_size >= 0 else -1
+        wpath = self._as_unc_path(path)
+        if autocommit is False:
+            # writes to temporary file, move on commit
+            # TODO: use transaction support in SMB protocol
+            share = path.split("/")[1]
+            path2 = "/{}{}/{}".format(share, self.temppath, uuid.uuid4())
+            wpath2 = self._as_unc_path(path2)
+            f = smbclient.open_file(wpath2, mode, buffering=buffering, **kwargs)
```

Comment:
Instead of patching the instance, you could make a wrapper class or subclass containing the extra methods

Reply:
Oh! I missed that.

I'll fix and commit next.

Reply:
Adressed in d8f676b

Reply:
Sorry I wasn't clear, I mean attaching methods to `f`, the file-like instance. It may be overkill, but you could do something like [this](https://github.com/intake/filesystem_spec/blob/master/fsspec/implementations/local.py#L189) for local files.

Reply:
I got it now!
Unfortunately my city was hit by strong winds and rain, and I don't have electricity to explore it right now.

Reply:
I hope this is addressed in e4d9988. 
Could you review again, please?